### DISCUSSION TITLE
is there Debugger for .NET

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#91805] [Q&A]
is there debugger for .NET, similar to how I can press F12 in browser to look at html

.NET reflection is similar to the concept of browser debugger right

**Comment by Clockwork-Muse** [comment]
If you mean "distributed with a built application", then no.  It's normally part of the dev tools/IDE (it's included in Visual Studio/VSCode), although you can use external debuggers as well.

Is there some specific problem you're trying to solve here?

**Reply by Xyncgas** [reply]
to have a debugger that's similar to browser's because how useful it is to debug webapp can be for .NET applications too, ideally debugging in way similar to looking at html element, it would be amazing if I can see in my web app the list of objects in the view model and what's in the object and so on

**Reply by Clockwork-Muse** [reply]
.... You mean a WYSIWYG (what you see is what you get) debugger?

That works on the web because everything has to write to the document (although it doesn't necessarily track as well for non-visual elements), but there's no common model for desktop/mobile applications.  That's because there are multiple display frameworks (or even "no framework" in some cases).

In theory the display framework could provide debugger integration that would allow something like this (and it would have to integrate with the IDE, since for a variety of reasons it wouldn't be in a shipping application), but there isn't really anything the runtime could provide.

That said, WPF does provide this with integration with Visual Studio, so it's available for at least one framework:
![image](https://github.com/dotnet/runtime/assets/5569508/ca2061a6-0423-4f5c-a2ba-f33d1005fc39)

It looks like Avalonia might as well.

**Reply by Xyncgas** [reply]
I wish visual studio would display a DOM like html in the debugger, that shows all the current object in my running program, and I can interact with these object by left clicking on them and use $0 in debugger, you know like you can in the browser's debugger
